packages/SwingSet/misc-tools/ingest-xs-snapshot.js

```diff
@@ -0,0 +1,744 @@
+      if (len > remaining) {
+        throw Error(`asked for ${len} from ${tag} but only ${remaining} left`);
+      }
+      const data = gather(len);
+      remaining -= len;
+      // console.log(` readN(${len}), remaining ${remaining+len} -> ${remaining}`);
+      return data;
+    };
+
+    const readByte = async () => {
+      const bytes = await readN(1);
+      return bytes[0];
+    };
+
+    const readS1 = async () => {
+      const bytes = await readN(1);
+      return new DataView(bytes.buffer).getInt8(0);
+    };
+
+    const readLE4 = async () => {
+      const bytes = await readN(4);
+      // the 'creation' (CREA) section is a raw dump of a C 'struct
+      // sx_creation', so it uses host byte order, and all computers
+      // are little-endian these days
+      return new DataView(bytes.buffer).getUint32(0, LITTLE_ENDIAN);
+    };
+
+    const readAll = () => readN(remaining);
+
+    async function* iterN(unitSize) {
+      while (remaining >= unitSize) {
+        yield readN(unitSize);
+      }
+      if (remaining !== 0) {
+        throw Error(
+          `done with ${tag} iterN(${unitSize}), but odd ${remaining} left`,
+        );
+      }
+    }
+
+    const skipSection = async () => {
+      if (!remaining) return;
+      await skipN(remaining);
+      remaining = 0;
+    };
+
+    const getRemaining = () => remaining;
+
+    return {
+      readN,
+      readByte,
+      readS1,
+      readLE4,
+      readAll,
+      iterN,
+      getRemaining,
+      skipSection,
+    };
+  };
+
+  return { readBE4, readTag, skipN, startSection };
+};
+
+async function readVersion(r) {
+  const s = await r.startSection('VERS');
+  const major = await s.readByte(1);
+  const minor = await s.readByte(1);
+  const patch = await s.readByte(1);
+  const endianValue = await s.readS1();
+  await s.skipSection();
+  let endian;
+  if (endianValue === 32) {
+    endian = 'le32'; // 32: size of txSlot
+  } else {
+    throw Error(`unhandled endian value: ${endianValue}`);
+  }
+  return { major, minor, patch, endian };
+}
+
+async function readCreation(r) {
+  const s = await r.startSection('CREA');
+  const initialChunkSize = await s.readLE4();
+  const incrementalChunkSize = await s.readLE4();
+  const initialHeapCount = await s.readLE4();
+  const incrementalHeapCount = await s.readLE4();
+  const stackCount = await s.readLE4();
+  const keyCount = await s.readLE4();
+  const nameModulo = await s.readLE4();
+  const symbolModulo = await s.readLE4();
+  const parserBufferSize = await s.readLE4();
+  const parserTableModulo = await s.readLE4();
+  const staticSize = await s.readLE4();
+  await s.skipSection();
+  return {
+    initialChunkSize,
+    incrementalChunkSize,
+    initialHeapCount,
+    incrementalHeapCount,
+    stackCount,
+    keyCount,
+    nameModulo,
+    symbolModulo,
+    parserBufferSize,
+    parserTableModulo,
+    staticSize,
+  };
+}
+
+// parsed from xsAll.h, look for XS_STRING_KIND
+const ALL_KINDS = [
```

Comment:
How much of this (and the rest of the ingestion logic) is going to be brittle in the face of changes to XS?  Should there be some kind of version check?

Reply:
good question.. I don't know. There are some version fields (major/minor/patch and an endian+wordsize field), but I don't think they're making any compatibility claims about them, at least none that outsiders could rely upon. So "yeah it's probably quite brittle", but I don't know that comparing the versions against anything would make it less so.

Reply:
Snapshots are supposed to be compatible within a major + minor range, and there's an explicit check failing the load if those differ.